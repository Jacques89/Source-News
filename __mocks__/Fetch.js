export function setupFetchStub(news) {
  return function fetchStub(_url) {
    return new Promise((resolve) => {
      resolve({
        json: () =>
          Promise.resolve({
            news,
          }),
      })
    })
  }
}