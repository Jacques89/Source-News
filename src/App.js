import React, { useState } from 'react'

import Navigation from './components/Navigation/Navigation'
import NewsContainer from './components/NewsContainer/NewsContainer'
import './App.scss'

const App = () => {
  return (
    <div>
      <Navigation />
      <NewsContainer />
    </div>
  )
}

export default App