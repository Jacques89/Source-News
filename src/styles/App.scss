body {
  background-color: rgb(155, 131, 100);
  background-color: rgb(9, 3, 3);
}
