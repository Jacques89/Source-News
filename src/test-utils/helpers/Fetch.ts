export function fetchStub(news: Object) {
  return function fetchStub(_url: RequestInfo) {
    return new Promise<any>((resolve) => {
      resolve({
        json: () =>
          Promise.resolve<any>({
            news
          })
      })
    })
  }
}
