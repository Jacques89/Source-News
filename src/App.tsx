import { useState, useEffect } from 'react'

import './styles/App.scss'

import Navigation from './features/Navigation/Navigation'
import NewsContainer from './pages/NewsPage/NewsPage'
import LoadingPage from './features/LoadingPage/LoadingPage'
import SearchInput from './features/SearchInput/SearchInput'
import Footer from './features/Footer/Footer'

require('dotenv').config()

type NewsProps = {
  category: string
  content: string
  img: string
  title: string
}

const App = () => {
  const [news, setNews] = useState<Array<NewsProps>>([])
  const [category, setCategory] = useState<string>('')
  const [loading, setLoading] = useState<boolean>(true)
  const [userInput, setUserInput] = useState<string>('')

  const newsAPI: string = `${process.env.REACT_APP_NEWS_API}`

  const dataFetch = async () => {
    try {
      // const request: Response = await fetch(newsAPI)
      // const response = await request.json()
      // setNews(response)
      setLoading(false)
      // return response
    } catch (error: unknown) {
      console.log('ERROR:', error)
    }
  }

  const changeCategory = (category: string) => {
    setCategory(category)
    setUserInput('')
  }
  const handleSubmit = (input: string) => {
    setUserInput(input)
  }

  useEffect(() => {
    dataFetch()
    setCategory('World')
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <div data-testid='app'>
      <Navigation changeCategory={changeCategory} category={category} />
      <SearchInput handleSubmit={handleSubmit} />
      {news.length > 1 ? (
        <NewsContainer news={news} category={category} userInput={userInput} />
      ) : null}
      {loading ? <LoadingPage /> : null}
      {news.length > 0 ? <Footer /> : null}
    </div>
  )
}

export default App
