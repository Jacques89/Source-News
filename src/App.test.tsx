import App from './App'
import {
  render,
  screen,
  act,
  waitFor,
  waitForElementToBeRemoved
} from '@testing-library/react'

import 'isomorphic-fetch'
import { setupFetchStub } from './test-utils/helpers/Fetch'

type NewsProps = {
  category: string
  content: string
  img: string
  title: string
}

describe('App', () => {
  it('renders correctly', async () => {
    act(() => {
      render(<App />)
    })
    await waitForElementToBeRemoved(() => screen.queryByText(/loading/i))
    expect(screen.getByTestId('app')).toMatchSnapshot()
  })
})

describe('API Request', () => {
  // TODO refactor
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(setupFetchStub(() => [{}]))
    // Clear
    const fetchSpy = jest.spyOn(global, 'fetch')
    fetchSpy.mockClear()
  })

  const mockNewsResponse: Array<NewsProps> = [
    {
      category: 'World',
      content: 'Squirrel becomes President',
      img: 'some Image URL',
      title: 'Squirrel has now become President of the World'
    },
    {
      category: 'Sports',
      content: 'Messi wins World Cup with Argentina',
      img: 'some url image',
      title: 'Messi finally wins a World Cup'
    }
  ]

  const url: string = `${process.env.REACT_APP_NEWS_API}`

  it('fetches data from API successfully', async () => {
    const fakeData: Array<NewsProps> = mockNewsResponse
    const fetchSpy = jest
      .spyOn(global, 'fetch')
      .mockImplementation(setupFetchStub(fakeData))

    const res: Response = await fetch(url)
    const json: Array<NewsProps> = await res.json()
    expect(json).toEqual({ news: fakeData })
    expect(fetchSpy).toHaveBeenCalledTimes(1)
  })
  it('returns an empty array if no data is available from API', async () => {
    const emptyData: Array<NewsProps> = []
    const fetchSpy = jest
      .spyOn(global, 'fetch')
      .mockImplementation(setupFetchStub(emptyData))

    const res: Response = await fetch(url)
    const json: Array<NewsProps> = await res.json()
    expect(json).toEqual({ news: emptyData })
    expect(fetchSpy).toHaveBeenCalledTimes(1)
  })
})
