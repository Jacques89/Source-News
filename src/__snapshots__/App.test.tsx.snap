// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  data-testid="app"
>
  <div
    data-testid="navigation"
    style="margin-top: 0px; transition-duration: .8s;"
  >
    <header>
      <h1>
        Source News
      </h1>
      <h3>
        Straight from the sources mouth!
      </h3>
      <svg
        class="icon-menu"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        />
      </svg>
    </header>
    <nav
      style="opacity: 0; display: none; z-index: 1;"
    >
      <ul
        aria-label="news-categories"
      >
        <li
          aria-label="category"
          style="background-color: rgb(255, 255, 255); color: rgb(0, 0, 0);"
        >
          World
        </li>
        <li
          aria-label="category"
          style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
        >
          Business
        </li>
        <li
          aria-label="category"
          style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
        >
          Sports
        </li>
        <li
          aria-label="category"
          style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
        >
          Entertainment
        </li>
        <li
          aria-label="category"
          style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
        >
          Technology
        </li>
      </ul>
    </nav>
  </div>
  <form
    action="#"
    class="search-bar"
    data-testid="searchInput"
    style="margin-top: 0px; transition-duration: .8s;"
  >
    <button
      class="search-btn"
    >
      <svg
        class="search-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        />
      </svg>
    </button>
    <input
      aria-labelledby="searchBar"
      placeholder="Search"
      type="text"
      value=""
    />
  </form>
  <div
    class="container"
  >
    <div
      class="box"
      style="opacity: 0; transition-delay: .5s; transition-duration: .5s;"
    >
      <article>
        <h3
          class="title"
        >
          Opening statements begin in Kyle Rittenhouse murder trial
        </h3>
        <div
          class="image"
          style="background-image: url(https://www.ctvnews.ca/polopoly_fs/1.5648480.1635867593!/httpImage/image.jpg_gen/derivatives/landscape_620/image.jpg);"
        />
        <p
          class="description"
        >
           KENOSHA, WIS.  --   Opening statements have begun in the murder trial of Kyle Rittenhouse, the aspiring police officer who shot three people who were out on the streets of Kenosha, Wisconsin, during a protest against racial injustice last year.
It took just one day to pick the jury Monday, despite the turbulent political passions unleashed by the shootings that left two dead and one wounded.
Rittenhouse was 17 when he traveled to Kenosha from his home in Illinois, just across the Wisconsin state line, in August 2020 after protests broke out over the shooting of a Black man by [...]
        </p>
        <p
          class="link"
        >
          <a
            href="https://www.ctvnews.ca/world/opening-statements-begin-in-kyle-rittenhouse-murder-trial-1.5648468"
          >
            Continue reading
          </a>
        </p>
      </article>
    </div>
  </div>
  <footer
    data-testid="footer"
  >
    <p
      class="copyright"
    >
      Jacques Nalletamby, 2021
    </p>
    <div
      class="icons"
    >
      <a
        class="icon"
        data-testid="githubLink"
        href="https://github.com/jacques89"
      >
        <svg
          data-testid="githubIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM277.3 415.7c-8.4 1.5-11.5-3.7-11.5-8 0-5.4.2-33 .2-55.3 0-15.6-5.2-25.5-11.3-30.7 37-4.1 76-9.2 76-73.1 0-18.2-6.5-27.3-17.1-39 1.7-4.3 7.4-22-1.7-45-13.9-4.3-45.7 17.9-45.7 17.9-13.2-3.7-27.5-5.6-41.6-5.6-14.1 0-28.4 1.9-41.6 5.6 0 0-31.8-22.2-45.7-17.9-9.1 22.9-3.5 40.6-1.7 45-10.6 11.7-15.6 20.8-15.6 39 0 63.6 37.3 69 74.3 73.1-4.8 4.3-9.1 11.7-10.6 22.3-9.5 4.3-33.8 11.7-48.3-13.9-9.1-15.8-25.5-17.1-25.5-17.1-16.2-.2-1.1 10.2-1.1 10.2 10.8 5 18.4 24.2 18.4 24.2 9.7 29.7 56.1 19.7 56.1 19.7 0 13.9.2 36.5.2 40.6 0 4.3-3 9.5-11.5 8-66-22.1-112.2-84.9-112.2-158.3 0-91.8 70.2-161.5 162-161.5S388 165.6 388 257.4c.1 73.4-44.7 136.3-110.7 158.3zm-98.1-61.1c-1.9.4-3.7-.4-3.9-1.7-.2-1.5 1.1-2.8 3-3.2 1.9-.2 3.7.6 3.9 1.9.3 1.3-1 2.6-3 3zm-9.5-.9c0 1.3-1.5 2.4-3.5 2.4-2.2.2-3.7-.9-3.7-2.4 0-1.3 1.5-2.4 3.5-2.4 1.9-.2 3.7.9 3.7 2.4zm-13.7-1.1c-.4 1.3-2.4 1.9-4.1 1.3-1.9-.4-3.2-1.9-2.8-3.2.4-1.3 2.4-1.9 4.1-1.5 2 .6 3.3 2.1 2.8 3.4zm-12.3-5.4c-.9 1.1-2.8.9-4.3-.6-1.5-1.3-1.9-3.2-.9-4.1.9-1.1 2.8-.9 4.3.6 1.3 1.3 1.8 3.3.9 4.1zm-9.1-9.1c-.9.6-2.6 0-3.7-1.5s-1.1-3.2 0-3.9c1.1-.9 2.8-.2 3.7 1.3 1.1 1.5 1.1 3.3 0 4.1zm-6.5-9.7c-.9.9-2.4.4-3.5-.6-1.1-1.3-1.3-2.8-.4-3.5.9-.9 2.4-.4 3.5.6 1.1 1.3 1.3 2.8.4 3.5zm-6.7-7.4c-.4.9-1.7 1.1-2.8.4-1.3-.6-1.9-1.7-1.5-2.6.4-.6 1.5-.9 2.8-.4 1.3.7 1.9 1.8 1.5 2.6z"
          />
        </svg>
      </a>
       
      <a
        class="icon"
        data-testid="linkedinLink"
        href="https://www.linkedin.com/in/jacquesnalletamby/"
      >
        <svg
          data-testid="linkedinIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
          />
        </svg>
      </a>
    </div>
  </footer>
</div>
`;
